src/python/pants/backend/python/macros/python_requirements.py

```diff
@@ -5,7 +5,7 @@
 from typing import Iterable, Mapping, Optional
 
 from pants.backend.python.target_types import parse_requirements_file
-from pants.base.build_environment import get_buildroot
+from pants.base.build_environment import get_buildroot, get_global_module_mapping
 
 
 class PythonRequirements:
```

Comment:
Rather than changing this macro (which is responsible for passing arguments into underlying `python_requirement_library` targets), you'd need to change https://github.com/pantsbuild/pants/blob/bd81ccaa96b343983d9bd62c7a514390a37e0cb4/src/python/pants/backend/python/dependency_inference/module_mapper.py#L177-L199 ... which is what consumes the `ModuleMapping` field.

Reply:
Agreed on not updating the macro, but I think the cleaner place to fix this is the `python_requirement_library` target itself. This avoids adding yet more complexity to the already complex module_mapper.py.

https://github.com/pantsbuild/pants/blob/824a15daa9f633554b5544ae2c59405601e20e1c/src/python/pants/backend/python/target_types.py#L586-L593

You'll override the `compute_value()` method like this: https://www.pantsbuild.org/docs/target-api-new-fields#adding-custom-validation. But rather than custom validation, you will mix-in the global default with something like

```python
provided_mapping = super().compute_value(raw_value, address=address) or {}
return {**DEFAULT_MODULE_MAPPING, **provided_mapping}
```

Two notes here:
1. `super().compute_value()` returns `dict | None`, and we need to handle that `None` case.
2. We want the user provided mapping to override the global default, the order of that dict unpacking matters.

--

While you're at it, please set `value: FrozenDict[str, Tuple[str, ...]]`, which will override the default of 

https://github.com/pantsbuild/pants/blob/824a15daa9f633554b5544ae2c59405601e20e1c/src/python/pants/engine/target.py#L1178-L1179

We know it will never be `None` anymore because you're guaranteed to mix in the default dictionary. This will allow you in the `module_mapper.py` code to remove this `or {}  # type: ignore`:

https://github.com/pantsbuild/pants/blob/bd81ccaa96b343983d9bd62c7a514390a37e0cb4/src/python/pants/backend/python/dependency_inference/module_mapper.py#L185